### PR TITLE
 Developer documentation ISIS SANS

### LINKED ISSUE #20335: Create developer documentation for ISIS SANS back-end
Provide developer documentation for the new ISIS SANS back-end. This documentation will focus on:
* Key features of the design
* the state approach
* reduction work-flow
* possible approaches to extending the framework to other instruments.

### PR DESCRIPTION
**Pull request by AntonPiccardoSelg**
Fixes #20335

This PR adds a developer documentation page for the ISIS SANS reduction framework.

**To test:**

Make sure that the documentation renders correctly.
Check for spelling mistakes

**Release Notes** 
Does not have to be in the release notes

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: docs/source/development/ISISSANSReductionBackend.rst [hunk 1]
```diff
@@ -0,0 +1,1221 @@
...
+
+This new implementation of the SANS data reduction uses a simple state object
+which stores the reduction-relevant information centrally and is immutable.
+This *SANSState* approach is the corner stone of the new design.
+The *SANSState* is:
+
+- self-validating
+- immmutable
+- typed
+- serializabvle
```
Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: serializable

File: docs/source/development/ISISSANSReductionBackend.rst [hunk 2]
```diff
@@ -0,0 +1,1221 @@
...
+The ``state_base.py`` module contains the essential ingredients for defining a
+state object. These are the ``StateBase`` class which allows for serialization
+and a set of ``TypedParameter``.
+
+The *StateBase*'s *property_manager* property is responsible for serialization.
+Due to the nature of the *PropertyManagerProperty* of algorithms it serializes
+the state object to a Python dictionary and receives a Mantid *PropertyManager*
+object.
+
+States which want to fulfil the *StateBase* contract must override the
```
Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: fulfill

File: docs/source/development/ISISSANSReductionBackend.rst [hunk 3]
```diff
@@ -0,0 +1,1221 @@
...
+
+
+Overview
+--------
+
+The reduction back-end consists of three components:
+
+- the *SANSState* approach to centrally store the state of the reduction
+- a set of work-flow algorithms which perform the individual reduction steps
+- a algorithm which orchestrates the work-flow algorithms.
```
Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: an algorithm

File: docs/source/development/ISISSANSReductionBackend.rst [hunk 4]
```diff
@@ -0,0 +1,1221 @@
...
+   f. Convert to wavelength and rebin using :ref:`SANSConvertToWavelengthAndRebin <algm-SANSConvertToWavelengthAndRebin>`
+
+5. Perform fitting for the transmission calculation. The sub-steps are:
+
+   a. Use incident monitor, wavelength settings, transmission detector ids
+      and fit settings to initialize :ref:`CalculateTransmission <algm-CalculateTransmission>`
+   b. Execute :ref:`CalculateTransmission <algm-CalculateTransmission>`
+   c. Get the fitted and unfitted workspace
+
+6. Set the fitted and unfitted workspaces o
```
Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: e.g.

Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: i.e. ?

File: docs/source/development/ISISSANSReductionBackend.rst [hunk 5]
```diff
@@ -0,0 +1,1221 @@
...
+object.
+
+States which want to fulfil the *StateBase* contract must override the
+*validate* method. This method is used to ensure internal consistency
+of the *TypedParameters* on the state. It is important to have comprehensive
+and tight checks here.
+
+The entries on the state objects are all of type *TypedParameter* which allows
+for type checking, ensuring consistency early on. It is easy to
+build custom types. The currently list of types are:
```
Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: current

File: docs/source/development/ISISSANSReductionBackend.rst [hunk 6]
```diff
@@ -0,0 +1,1221 @@
...
+*reduction_mode.py*
+*******************
+
+The *StateReductionMode* class contains general settings about the reduction, e.g. if we are dealing with a merged
+reduction. It contains the following parameters:
+
+=============================== =================================================== ============================================== ========= =============== ===========================================
+Name                            Comment                                             Type                                           Optional? Auto-generated? Default value
+=============================== =================================================== ============================================== ========= =============== ===========================================
+reduction_mode                  The type of reduction, ie LAB, HAB, merged or both  *ClassTypeParameter(ReductionMode)*            N         N               *ISISReductionMode.LAB* enum value
```
Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: ie ?

File: docs/source/development/ISISSANSReductionBackend.rst [hunk 7]
```diff
@@ -0,0 +1,1221 @@
...
+merge_range_max                 The max q value for merging                         *FloatWithNoneParameter*                       Y         N               *None*
+detector_names                  A dict from detector type to detector name          *DictParameter*                                N         Y               -
+=============================== =================================================== ============================================== ========= =============== ===========================================
+
+
+*slice.py*
+**********
+
+The *StateSliceEvent* class is only relevant when we are dealing with event-type
+data and the user decides to perform an event-sliced reduction, ie one reduction per event slice.
```
Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: ie ?

File: docs/source/development/ISISSANSReductionBackend.rst [hunk 8]
```diff
@@ -0,0 +1,1221 @@
...
+
+
+*adjustment.py*
+***************
+
+Adjustment workspaces are generated to be consumed in the momentum transfer conversion step.
+There are three types of adjustments
+
+- Pure wavelength adjustments, ie adjustments which only affect the bins
+- Pure pixel adjustments, ie adjustments which only affect the spectra
```
Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: ie ?

File: docs/source/development/ISISSANSReductionBackend.rst [hunk 9]
```diff
@@ -0,0 +1,1221 @@
...
+- Pixel-and-wavelength adjustments, ie adjustments which affect both the bins and spectra
```
Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: ie ? ...

File: docs/source/development/ISISSANSReductionBackend.rst [hunk 10]
```diff
@@ -0,0 +1,1221 @@
...
+for the circular or the rectangular cases needs to be specified.
+
+
+*compatibility.py*
+******************
+
+The *StateCompatibility* class is not directly part of the reduction, but it will
+convert event-mode workspaces early on to histogram-mode workspaces in order to
+emulate the old reduction work-flow. This allows for a direct comparison between
+results of the new and old reduction framework. The name compatibility mode has
```
Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: ~~mode~~

File: docs/source/development/ISISSANSReductionBackend.rst [hunk 11]
```diff
@@ -0,0 +1,1221 @@
...
+which means that the director which uses these settings will use the last value in the list.
+
+The user input dictionary is normally populated by the settings specified in the user file and
+which are parsed by *UserFileParser* in *user_file_parser.py*. In additin the dictinary can
+be modified by using the *ISISCommandInterface* or the SANS GUI. Changes to the original
+settings will override settings specified in the user file.
+
+
+An example dictionary entry for the fit parameters during the transmission
+calculation for a Can data set could be:
```
Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: can

File: docs/source/development/ISISSANSReductionBackend.rst [hunk 12]
```diff
@@ -0,0 +1,1221 @@
...
+
+.. code-block:: python
+
+  {FitId.general: fit_general(start=1.0,
+                              stop=3.0,
+                              fit_type=FitType.Polynomial,
+                              data_type=DataType.Can,
+                              polynomial_order=2)
+
+This entry is added to the general user input dictionary. Note that for some ofthe
```
Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: of the

File: docs/source/development/ISISSANSReductionBackend.rst [hunk 13]
```diff
@@ -0,0 +1,1221 @@
...
+The main director which handles the coordination of the builders and the only one which
+is actually aware of them is *StateDirectorISIS*. The director manages the user input
+dictionary which was discussed earlier. It is also possible to provide a user file as input or
+a user input dictionary.
+
+This director is used by other directors which are responsible for creating the user input
+dictionary for the CLI and GUI case. These directors don't know anything about the builders or
+the state, but are only responsible for providing the user input. An exception to this
+is the *StateData* object, since it is used indirectly to choose the correct builders
+for the other sub states. Hence the role of the outer level directors is to provide the
```
Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: sub-states

File: docs/source/development/ISISSANSReductionBackend.rst [hunk 14]
```diff
@@ -0,0 +1,1212 @@
...
+   b. Execute :ref:`CalculateTransmission <algm-CalculateTransmission>`
+   c. Get the fitted and unfitted workspace
+
+6. Set the fitted and unfitted workspaces on the output of the algorithm.
+
+
+``SANSConvertToQ``
+------------------
+
+The ``SANSConvertToQ`` algorithm is the
```
Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: N \cross M  maybe \times works in this case (cross product)?

I get:

System Message: WARNING/2 (N \cross M)

latex exited with error [stdout] This is pdfTeX, ...

Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: same here.

File: docs/source/development/ISISSANSReductionBackend.rst [hunk 15]
```diff
@@ -1202,8 +1203,63 @@ to perform the reduction.
...
+inner core of the orchestration mechanism. The inputs to this algorithm are
+
+- A SANS state object
+- Several input workspaces
+- A selection which detector type we are dealing with, i.e. *LAB* or *HAB*
+- A selection which data type we are dealing with, i.e. *Sample* or *Can*
+
+The sub-steps of this algorithm are:
+
+1. Get the cropped input *ScatterWorkspace*. The cropping is defined by the selected dector type.
```
Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: detector

File: docs/source/development/ISISSANSReductionBackend.rst [hunk 16]
```diff
@@ -1159,7 +1160,7 @@ reduction, but rather only reduces either the sample or the can dataself.
 For this the ``SANSSingleReduction`` algorithm was developed. It runs ``SANSReductionCore``
 with the appropriate data (sample or can) and performs the required post processing, e.g. stitching.
 This algorithm will produce a fully reduced output. However it will not produce it in the desired form,
-eg correct name of the output workspaces, grouping of mult-period reduced data etc. This is achieved with
+e.g. correct name of the output workspaces, grouping of mult-period reduced data etc. This is achieved with
```
Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: multi-period

File: docs/source/development/ISISSANSReductionBackend.rst [hunk 17]
```diff
@@ -1202,8 +1203,63 @@ to perform the reduction.
...
+   This uses the input *TransmissionWorkspace* and *DirectWorkspace* workspaces. Note that
+   they are moved using :ref:`SANSMove <algm-SANSMove>` before they are used by the adjustment
+   algorithm. The outputs are a wavelength-adjustment workspace, a pixel-adjustment worspace and a wavelength-and-pixel adjustment
+   workspace. Note that the creation is optional.
+9. Convert the data workspace into histogram-mode using :ref:`RebinToWorkspace <algm-RebinToWorkspace>`.
+   This is only relevant for event-mode workspaces where the compatibility mode has not been used. Up until
+   now the event-mode workspace could be used as an event workspace, but the momentum transfer conversion (the next step)
+   requires a histogram-mode workspace.
+10. The final step, the conversion to momentum transfer units, either uses :ref:`Q1D <algm-Q1D>`
+    or :ref:`Qxy <algm-Qxy>` depending what the setting of the reduction dimensionality is. This step
```
Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: depending on?

**Comment by AntonPiccardoSelg** [comment]
I am placing this back into `3.11`, especially since it is entirely new documentation with very little risk

File: docs/source/development/ISISSANSReductionBackend.rst [hunk 18]
```diff
@@ -0,0 +1,1290 @@
...
+wavelength_step_type  This is either linear or logarithmic *ClassTypeParameter(RangeStepType)* N         N
+===================== ==================================== =================================== ========= ===============
+
+The validation ensures that all entries are specified and that the lower wavelength boundary is smaller than the upper wavelength boundary.
+
+*save.py*
+*********
+
+The *StateSave* class does not hold information which is directly related to the reduction but contains
+the all the required information about saving the reduced data. The relevant parameters are:
```
Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: ~~the~~ all ~~the~~ required information

File: docs/source/development/ISISSANSReductionBackend.rst [hunk 19]
```diff
@@ -0,0 +1,1290 @@
...
+background_TOF_roi_start         Lower bound of background correction when using ROI on detector                                  *FloatParameter*                Y         N               -
+background_TOF_roi_stop          Upper bound of background correction when using ROI on detector                                  *FloatParameter*                Y         N               -
+fit                              A dict for each data type (sample and can) to the state of fit settings (*StateTransmissionFit*) *DictParameter*                 Y         N               -
+================================ ================================================================================================ =============================== ========= =============== =======
+
+Note that the transmission information can be either collected via a monitor or
+via a region on the detector. In the former case *transmission_monitor* is the
+relevant parameter whereas in the former case it is *transmission_radius_on_detector*,
```
Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: latter case?

File: docs/source/development/ISISSANSReductionBackend.rst [hunk 20]
```diff
@@ -0,0 +1,1290 @@
...
+These settings are captured in an input dictionary and then processed
+by builder classes which are coordinated by a state director. These components are
+described below.
+
+User input dictionary
+^^^^^^^^^^^^^^^^^^^^^^
+
+As mentioned above, we cannot make any assumptions about the order or multiplicity of the
+user commands. We use a simple Python dictionary to store the specified settings.
+In fact the dictionary maps from enum-like classes, defined in *settings_tags.py* to
```
Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: In fact, ?

File: docs/source/development/ISISSANSReductionBackend.rst [hunk 21]
```diff
@@ -0,0 +1,1290 @@
...
+a list of settings. The settings can be simple values, lists, dictionaries or *named_tuples* defined
+in *settings_tags.py*.
+
+Note that the naming of a large chunk of the the enum-like classes in *settings_tags.py*
+was driven by the corresponding name in the user file definition. We can consider
+changing the naming in the future. Also note that some settings only allow one value,
+which means that the director which uses these settings will use the last value in the list.
+
+The user input dictionary is normally populated by the settings specified in the user file and
+which are parsed by *UserFileParser* in *user_file_parser.py*. In addition the dictionary can
```
Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: ~~which are~~

File: docs/source/development/ISISSANSReductionBackend.rst [hunk 22]
```diff
@@ -0,0 +1,1290 @@
...
+   a. Use incident monitor, wavelength settings, transmission detector ids, fit
+      settings as well as the corrected transmission (step 3) and direct (step4)
+      workspaces to initialize :ref:`CalculateTransmission <algm-CalculateTransmission>`
+   b. Execute :ref:`CalculateTransmission <algm-CalculateTransmission>`
+   c. Get the fitted and unfitted output workspaces
+
+6. Set the fitted and unfitted workspaces on the output of the algorithm.
+
+
+*SANSCo
```
Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: moves a SANS instrument ?

Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: ~~which~~ ? As well for the following 2 lines?

Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: used later on in ?? I guess there is a name missing

Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: workspaces are moved? see earlier comment

Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: their creation?

Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: ~~what~~ ... ~~is~~ ?

Comment (reimundILL) [inline docs/source/development/ISISSANSReductionBackend.rst]: I would delete that line. There are no Footnotes!

**Review by vreimund — APPROVED** [review]
Congratulation to this documentation for developers!
Documentation renders well for me.
In general, category Development cannot be found very quickly (I used the search field in the MantidPlot help menu).
I hope this documentation stays up-to-date in the future even when SANS reduction design will experience updates.
Maybe I would add (later) a block diagram to visualize dependent algorithms.
English was not my first foreign language (but Latin).

**Comment by AntonPiccardoSelg** [comment]
Hi Verena, 
Thanks for your comments, they are very helpful. 

The last seven comments seem to all be placed against the same location in the file. That does not seem quite right. Would you mind placing them at the location where you intended them to be. Again, they are very helpful and I would like to provide corrections for these suggestions.

**Comment by martyngigg** [comment]
I'm not going to wait for the system test build as it's just a documentation change.

**Pull request merged**